src/chunk.js

```diff
@@ -0,0 +1,181 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from './log';
+import {fromClassForDoc} from './service';
+import {isExperimentOn} from './experiments';
+import {makeBodyVisible} from './style-installer';
+import {viewerPromiseForDoc} from './viewer';
+
+/**
+ * @const {boolean}
+ */
+const shouldNotUseMacroTask = /nochunking/.test(self.location.href);
+
+/**
+ * @const {!Promise}
+ */
+const resolved = Promise.resolve();
+
+/**
+ * Run the given function. For visible documents the function will be
+ * called in a micro task (Essentially ASAP). If the document is
+ * not visible, tasks will yield to the event loop (to give the browser
+ * time to do other things) and may even be further delayed until
+ * there is time.
+ *
+ * @param {!Node|!./service/ampdoc-impl.AmpDoc} nodeOrAmpDoc
+ * @param {function()} fn Function that will be called as a "chunk".
+ */
+export function chunk(nodeOrAmpDoc, fn) {
+  const service = fromClassForDoc(nodeOrAmpDoc, 'chunk', Chunk);
+  service.run_(fn);
+};
+
+/**
+ * Runs all currently scheduled chunks.
+ * @param {!Node|!./service/ampdoc-impl.AmpDoc} nodeOrAmpDoc
+ */
+export function runChunksForTesting(nodeOrAmpDoc) {
+  const service = fromClassForDoc(nodeOrAmpDoc, 'chunk', Chunk);
+  const errors = [];
+  while (true) {
+    try {
+      if (!service.execute_()) {
+        if (errors.length) {
+          break;
+        }
+        return;
+      }
+    } catch (e) {
+      errors.push(e);
+    }
+  }
+  throw errors[0];
+}
+
+class Chunk {
+  /**
+   * @param {!./service/ampdoc-impl.AmpDoc} ampDoc
+   */
+  constructor(ampDoc) {
+    /** @private @const */
+    this.ampDoc_ = ampDoc;
+    /** @private @const {!Window} */
+    this.win_ = ampDoc.win;
+    /** @private @const {!Array<function()>} */
+    this.tasks_ = [];
+    /** @private {?./service/viewer-impl.Viewer} */
+    this.viewer_ = null;
+    /** @private @const {function()} */
+    this.boundExecute_ = () => this.execute_();
+    /** @private @const {boolean} */
+    this.active_ = isExperimentOn(this.win_, 'chunked-amp')
+        && !shouldNotUseMacroTask;
+
+    if (!this.active_) {
+      return;
+    }
+    if (!this.win_.requestIdleCallback) {
+      this.win_.addEventListener('message', this.boundExecute_);
+    }
+    viewerPromiseForDoc(ampDoc).then(viewer => {
+      this.viewer_ = viewer;
+      viewer.onVisibilityChanged(() => {
+        if (viewer.isVisible()) {
+          this.execute_();
+        }
+      });
+      if (viewer.isVisible()) {
+        this.execute_();
+      }
+    });
+  }
+
+  /**
+   * Run fn as a "chunk". It'll run in a micro task when the doc is visible
+   * and otherwise run it after having yielded to the event queue once.
+   * @param {function()} fn
+   */
+  run_(fn) {
+    this.tasks_.push(fn);
+    this.schedule_();
```

Comment:
Should "schedule next" be done from the starting point of `run_` here, or from the point when previous task was executed in `execute_`?